### PR TITLE
Add fastpath for cloning fully-buffered BlobCacheBufferedIndexInput

### PR DESCRIPTION
**Pull request by original-brownbear**
Same as [#107461](https://github.com/elastic/elasticsearch/pull/106794)

Like with slicing we can optimize cloning the same way. This has lots of potential for saving memory when dealing with offheap fst stores and the like where there's a bunch of small slices getting cloned.

Inspecting heap dumps suggests this might be a O(10M) win in at least one of our obs. clusters.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/lucene/store/ByteArrayIndexInput.java [hunk 1]
```diff
@@ -67,22 +67,22 @@ public long length() {
 
     @Override
     public byte readByte(long pos) throws IOException {
-        return bytes[position(pos)];
+        return bytes[offset + position(pos)];
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/common/lucene/store/ByteArrayIndexInput.java:70]: Would it be possible to have the bugfix parts here extracted to a separate PR?

**Review by henningandersen — COMMENTED** [review]
I think the whole PR is good, but I suppose Adrien may have a bit more expertise on how lucene uses clone. However, I'd like to expedite the bugfix part of this PR if possible, can we extract that part to a separate PR while we wait for his review?

File: server/src/main/java/org/elasticsearch/common/lucene/store/ByteArrayIndexInput.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/common/lucene/store/ByteArrayIndexInput.java:70]: I'm on it sorry for mixing these two in the first place, couple min! :)

Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/common/lucene/store/ByteArrayIndexInput.java:70]: Thanks!

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/BlobCacheBufferedIndexInput.java [hunk 2]
```diff
@@ -352,6 +353,23 @@ protected final IndexInput trySliceBuffer(String name, long sliceOffset, long sl
         return null;
     }
 
+    @Nullable
+    protected final IndexInput tryCloneBuffer() {
+        if (buffer.limit() == length && bufferStart == 0) {
```
Comment (jpountz) [inline x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/BlobCacheBufferedIndexInput.java:358]: Isn't this check already covered by the range check in trySliceBuffer?

**Review by jpountz — APPROVED** [review]
I left a question but in general LGTM, it makes sense to me that this sort of thing can help.

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/BlobCacheBufferedIndexInput.java (hunk 2 above)
Comment (original-brownbear) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/BlobCacheBufferedIndexInput.java:358]: Sort of, the `0` start offset isn't and we need it. I figured it's a little easier to read this by adding the other check and it saves cycles for range object allocations (maybe). But could probably do without the first condition indeed, I'll leave it for now hope that's ok.

**Comment by original-brownbear** [comment]
Thanks Adrien!
